File: python/leetcode/450. Delete Node in a BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def deleteNode(self, root: TreeNode, key: int) -> TreeNode:
        if root is None:
            return None

        if key > root.val:
            root.right = self.deleteNode(root.right, key)
        elif key < root.val:
            root.left = self.deleteNode(root.left, key)
        else:
            if root.left is None and root.right is None:
                root = None
            elif root.left is not None and root.right is not None:
                min = self.small(root.right)
                root.val = min.val
                root.right = self.deleteNode(root.right, root.val)
            else:
                if root.right is None:
                    root = root.left
                else:
                    root = root.right
        return root

    def small(self, root):
        while root.left is not None:
            root = root.left
        return root

File: c++/leetcode/450. Delete Node in a BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
  TreeNode *small(TreeNode *);

public:
  TreeNode *deleteNode(TreeNode *root, int key)
  {
    if (!root)
      return root;

    if (key > root->val)
      root->right = deleteNode(root->right, key);
    else if (key < root->val)
      root->left = deleteNode(root->left, key);
    else
    {
      // found
      if (!root->right && !root->left)
      {
        delete root;
        root = nullptr;
      }
      else if (root->right && root->left)
      {
        TreeNode *sm = small(root->right);
        root->val = sm->val;
        root->right = deleteNode(root->right, root->val);
      }
      else
      {
        if (!root->right)
          return root->left;
        else
          return root->right;
      }
    }

    return root;
  }
};

TreeNode *Solution::small(TreeNode *root)
{
  while (root->left)
    root = root->left;
  return root;
}